### ISSUE TITLE
Allow preprocessor plugins to work with the AST

### ISSUE THREAD

**Issue by tqc** [eslint/eslint#2610]
A preprocessor plugin should be able to produce an AST directly rather than just stripping out anything that isn't javascript.

My initial use case for this is linting of angular templates. It should be fairly easy to convert 

    Text content {{variable}} more text

into the AST for

    "Text content " + ngbind(variable) + " more text"

From there, writing rules for things like disallowing calculations in bound expressions is standard ESLint - all I need is something to connect it to. 

Changes required:

1. Split api.verify. See #1025, but I think this is simpler than setting up a custom parameter type.
 - api.verify(text, ...) same as existing function
 - api.verifyAST(ast, text, ...) called by api.verify after parse step

2. In cli-engine processFile, choose the appropriate variation of verify to call based on whether the preprocessor returned strings or an AST. Most of the concerns on #1025 seem to be around verifying that the AST is valid when it can come from anywhere. I'm not sure if verifyAST could be exposed to processFile without being fully public, but if it is for internal use only through the plugin system, anything wrong with the AST is a bug in the plugin, not core, and the validation requirement goes away.

3. Pass additional parameters to processor.preprocess exposing the default parser. This could be as a reference to the linter instance or just wrapping parse(text, config) so that it doesn't require the config parameter. This would allow the output of the preprocessor to be something like 

```
     CustomHtmlParser("<p>Text content</p>")
     + parse("< script>x=1;</script>".replace("< script>",""))
     + CustomHtmlParser("<p>More text content</p>")'
```

without the plugin needing to know anything about the default js parser or its options.

**Comment by ilyavolodin** [comment]
Being able to pass AST for validation has been discussed multiple times: #1025 As mentioned in that issue, AST alone is not enough, we need all tokens as well as text of the file itself. Many rules (like any white space rule) rely on tokens and text. And since you will have to generate text, you might as well pass it to `verify` and have it parse it into AST.

**Comment by tqc** [comment]
When I say AST, I mean the output from parse() - an object including tokens. Maybe not technically an AST, but a lot easier to refer to. Being a plugin we also have the original text, though it may be text that the standard parser wouldn't like. 

I did look at generated JS code and linting that, but then the error messages would refer to generated code rather than the original source, making the whole exercise somewhat pointless. Stripping out the unparseable bits and returning a string array has similar issues.

The basic problem is that my plugin knows things about the original source that verify(text) can't. As an overly simplified example:

```
<p>Text</p>
<script>
x=1
</script>
<p>More Text</p>
```

If I return ["x=1"] from the plugin, verify(text) will tell me there is an error in line 1. If I have access to the AST, I can update all the loc/range values so that the rules see that code correctly as line 3.

Some rules may not work where the original text is different to what is expected, but those would likely not make sense in this context and can safely be suppressed by the plugin, either by inserting a comment or removing the messages in the post processor - another reason that allowing a plugin to provide an AST is simpler than allowing it generally.

That said, I don't think incompatible rules is likely to be a huge problem - in the core rules there are only 9 making to getSource(), none of which seem to care much what the parser thinks of the text. The other obvious candidates like indent are based on AST position information.

**Comment by ilyavolodin** [comment]
Eee... I think you might've missed big part of what processors provide. There are two functions that a processor supplies to ESLint, `preprocess` and `postprocess`. A good processor, given you example above should return error on line 3, not line 1, that's the whole point of postprocess. First eslint pipes code through preprocess, that's supposed to get the code into parsable state, it also supposed to create a map of a sort, that would indicate difference in position between original file, and result. Once ESLint verifies supplied code, postprocess will be called with the list of errors, which postprocess should convert to actual locations, using the map created earlier.

While there are only 9 rules that use source, most of them run regular expression on the whole source, so you will most likely always get errors in those rules, if you give it a source code of HTML file (or Angular template). In general, I don't have a problem with preprocess outputting AST + Tokens, but #1025 will have to be done first to allow for that, and it's a low priority issue for us currently.

**Comment by ilyavolodin** [comment]
Btw, here's an example of processor that maps error back to the original location in the code: https://github.com/BenoitZugmeyer/eslint-plugin-html/blob/master/src/index.js
If you look at postprocess function, you can see that it uses source map to find location of the error in the original source code (html).

**Comment by tqc** [comment]
Agree that the overly simplified example could be implemented with mapping in postprocess. However, what I am actually looking to do is significantly more complex than I can fit in a three line example. 

Most of the regular expressions I see are just looking for newlines - the whole reason they look at the text rather than the AST is that the code is irrelevant for that rule.

I could do a PR for #1025, but the hard part of that is SourceCode.prototype.validate, which isn't really needed for this. Essentially I'm proposing a partial fix for #1025 - Allow the plugin to pass in anything it wants to, with the logic that someone writing a plugin should know more about ESLint internal requirements than someone just calling verify with a random AST. This would be very similar to the existing situation with parsers - you can specify any parser you want in .eslintrc, but if it doesn't behave exactly like esprima there is no guarantee it will work.

**Comment by ilyavolodin** [comment]
Let me ask this, if your processor will create a custom AST from a file and pass it to verify, and that custom AST will be in the ESTree format, why not create a custom parser all on it's own, that can be plugged into ESLint? From what you are describing, you will basically implement that in a processor, but ESLint already provides a way to use custom parser. It seems like a duplication of already existing functionality. 
Granted, we do not currently provide a way to specify parser (or configuration) per file extension, but that's been discussed as one of the features for version 2.0.0.
Would that solve the problem?

**Comment by tqc** [comment]
I considered that, but it didn't seem the right way to go when there is a plugin system. The basic parser would be something like

```
function parse(text) {
    if (extension == "html") return htmlparser.parse(text)
    else return esprima.parse(text)
}
```
That much is simple enough, but now how does it know that it should actually be using esprima-fb? What if I want to include more than one custom parser in a project? It works, but making it robust quickly ends up reimplementing a lot of stuff that is already there in the core.

Also, ESLint doesn't tell the parser what the original filename is, so I wouldn't be able to accept html tags in html files without also accepting them in js files, which is wrong.

Among other things, my approach would effectively let you specify a parser for a specific file extension without requiring any changes to the config file format - so it wouldn't be a breaking change that has to wait for a new major version.  

While you could put a processor on .js that just overrides the parser setting, there wouldn't be much point in it, since that plugin could be replaced with a single line in .eslintrc. Config for the basic case, plugins for something more complex.

**Comment by ilyavolodin** [comment]
I think you are over complicating it a bit. Here's how I would do this (until config per file extension are added).
Create a parser that only handles HTML files, no JS/JSX/Etc.
Create an empty processor and mark it for HTML files. You don't have to publish it, just npm link it to your project.
Create two grunt/gulp/webpack/whetere you are using for a build system, one for JS file, and one for HTML files with separate configs, and in HTML config specify your own parser.
While I do understand what you are saying, and I defiantly can see that what I'm proposing is not very pretty, but I think most of those issues are going to be solved once we add ability to provide separate configurations (including parser configuration) per file extension. I think what you are looking for is sort of outside of the scope of what processors where built for.

**Comment by tqc** [comment]
Pretty sure I'm simplifying things actually :)

Plugins become significantly more powerful, allowing a lot of stuff that would otherwise require adding complexity to core.

I don't think running two separate configs would work very well. Maybe if I was just running ESLint as a build step, but things like linting in the editor would require rewriting more of the toolchain than I am comfortable with to pick the right config.

Is there an ETA on the multiple configurations? From what I can see it's just a comment on a 2.0 wishlist, which makes it reasonable to consider better ways of doing it, especially if it can be done without major changes.  

If you don't think it fits into the processor, would you be more open to a new plugin feature, say plugin.parse? Same general idea, but leaving the processor functions unchanged.  Another option would be to have a function allowing a plugin to override configuration for the current file. That way the config file wouldn't need to do glob matching because the plugin already has that capability, and using a different file extension most likely means needing plugin anyway. That may require adding a plugin config section to .eslintrc to do properly, but that seems like it would be both simpler and more flexible than splitting .eslintrc by extension.

**Comment by ilyavolodin** [comment]
There's no ETA on multiple configurations currently. It hasn't been approved or discussed yet, but there are enough people who want it, that I don't doubt that it will make it into ESLint at some point in the future.
You are correct about two configs and linting in IDE:-(
I think I really would want to hear what @nzakas thinks about your ideas before accepting and committing to making this happen. It touches on a lot of core architecture of ESLint and plugins.

**Comment by nzakas** [comment]
To start, your initial proposal is too complex - I don't think its a good idea based on the number of areas that would have to change. 

We also don't deal just with ASTs, we need tokens, source text, and comments - so one way or another we'd need the JS as a string, which is why processors work with text and not ASTs. 

At the moment, I'm having trouble understanding why processors as they are today don't solve your use case.

**Comment by tqc** [comment]
Ok, I think the initial idea can be simplified a bit. 1 goes away if #1025 is handled first, which looks like being easier to do than I initially thought. 3 can probably be handled separately - it would definitely be nice to have access to the context/config from a processor, and it ties into some of my ideas around multiple configuration, but it doesn't directly block the others.

That leaves just the processor/verify call itself, which is suitably isolated.

Having direct access to the AST should make the plugin code needed for some of my more advanced ideas a lot simpler. 

```
custom parser -> verify -> print messages 
```

is a lot simpler and less error-prone than

```
custom parser -> complex mapping -> generated js -> default parser -> verify -> messages -> complex mapping -> print messages
```

**Comment by nzakas** [comment]
I'm sorry, I don't see us moving forward with this idea. You're asking for changes that are too dramatic for a use case that is uncommon. My advice remains: do text-to-text transform in the processor.
